backends/dpdk/dpdkProgram.cpp

```diff
@@ -187,6 +187,75 @@ IR::Declaration_Variable *ConvertToDpdkParser::addNewTmpVarToMetadata(cstring na
+void ConvertToDpdkParser::getCondVars(const IR::Expression *sv, const IR::Expression *ce,
+                                      IR::Expression **leftExpr, IR::Expression **rightExpr) {
+    if (auto maskexpr = ce->to<IR::Mask>()) {
+        auto left = maskexpr->left;
+        auto right = maskexpr->right;
+        /* Dpdk architecture requires that both operands of &&& be constants */
+        if (!left->is<IR::Constant>() || !left->is<IR::Constant>()) {
+            ::error(ErrorType::ERR_UNSUPPORTED,
+                    "Non constant values are not supported in Mask operation");
+            return;
+        }
+        unsigned value = right->to<IR::Constant>()->asUnsigned() &
+                         left->to<IR::Constant>()->asUnsigned();
+        auto tmpDecl = addNewTmpVarToMetadata("tmpMask", sv->type);
+        auto tmpMask = new IR::PathExpression(
+                           IR::ID("m." + tmpDecl->name.name));
+        collector->push_variable(new IR::DpdkDeclaration(tmpDecl));
+        add_instr(new IR::DpdkMovStatement(tmpMask, sv));
+        add_instr(new IR::DpdkAndStatement(tmpMask, tmpMask, right));
+        *leftExpr = tmpMask;
+        *rightExpr = new IR::Constant(value);
+    } else {
+        *leftExpr = sv->clone();
+        *rightExpr = ce->clone();
+    }
+}
+
+/* This is a helper function for handling tuple expressions in select statement in parser block.
+   The tuple expressions are evaluated as exact match of each element in the tuple and is
+   translated to dpdk assembly by short-circuiting the conditions where beginning of each keyset
+   starts with a unique case label.
+
+   transition select(a,b) {              Equivalent dpdk assembly:
+       (a1, b1) : state1;
+       (a2, b2) : state2;                jmpneq case1 a, a1
+       (a3, b3) : state3;                jmpneq case1 b, b1
+       }                    ==>          jmp state1
+   is evaluated as                       case1:   jmpneq case2 a, a2
+   if (a == a1 && b == b1)               jmpneq case2 b, b2
+       goto state1                       jmp state2
+   else (a == a2 && b == b2)             case2:    jmpneq case a, a3
+       goto state2                       jmpneq case2 b, b3
+   .....                                 jmp state3
+                                         ....
+    This function emits a series of jmps for each keyset.
+*/
+void ConvertToDpdkParser::handleTupleExpression(const IR::ListExpression *cl,
+                                                const IR::ListExpression *input, int inputSize,
+                                                cstring trueLabel, cstring falseLabel) {
+    IR::Expression *left;
+    IR::Expression *right;
+    /* Compare each element in the input tuple with the keyset tuple */
+    for (auto i = 0; i < inputSize; i++) {
+        auto switch_var = input->components.at(i);
+        auto caseExpr = cl->components.at(i);
+        if (!caseExpr->is<IR::DefaultExpression>()) {
```

Comment:
A simpler solution may be to treat each DefaultExpression like 0 &&& 0.  See the SingleElementSelect pass too.

Reply:
> If dpdk is limited to 64 bits per value this implementation is more general

Yes, header and metadata field width is limited to 64-bit and immediate values width is limited to  32-bit.

Reply:
> A simpler solution may be to treat each DefaultExpression like 0 &&& 0. See the SingleElementSelect pass too.

I am not really sure how converting default expr to 0 &&& 0 would be any better than current code. This code would generate a jmpneq for 0 &&& 0 (which would be false always and the jmpneq would I think be optimized away later) but the current approach would not generate any such jmpneq. Am I missing something here?

Reply:
In the implementation mentioned the total width of the select expression is limited to whatever the code generator supports.

Reply:
It's not better but there is one fewer case to handle.

Reply:
I think we can give an error till we do this transformation.